### PR TITLE
NEW add Ticket tab on contract

### PR DESCRIPTION
**Pull request by defrance**
Allow to add contract link on ticket
Allow to show associated ticket to a  contact on a tab

### CODE REVIEW

File: htdocs/ticket/card.php [hunk 1]
```diff
@@ -805,7 +806,7 @@
 
 	print '</form>'; */
 } elseif (empty($action) || $action == 'view' || $action == 'addlink' || $action == 'dellink' || $action == 'presend' || $action == 'presend_addmessage' || $action == 'close' || $action == 'abandon' || $action == 'delete' || $action == 'editcustomer' || $action == 'progression' || $action == 'categories' || $action == 'reopen'
-	|| $action == 'editsubject' || $action == 'edit_extras' || $action == 'update_extras' || $action == 'edit_extrafields' || $action == 'set_extrafields' || $action == 'classify' || $action == 'sel_contract' || $action == 'edit_message_init' || $action == 'set_status' || $action == 'dellink') {
+	|| $action== 'contrat' || $action == 'editsubject' || $action == 'edit_extras' || $action == 'update_extras' || $action == 'edit_extrafields' || $action == 'set_extrafields' || $action == 'classify' || $action == 'sel_contract' || $action == 'edit_message_init' || $action == 'set_status' || $action == 'dellink') {
```
Comment (eldy) [inline htdocs/ticket/card.php:809]: We try to standardize value for $action so we can use it for security purpose, by having a code starting with add, del, set, edit, ....

Can you rename the action = 'contrat'  into  'setcontract'  ?

File: htdocs/ticket/card.php [hunk 2]
```diff
@@ -1000,6 +1001,33 @@
...
 
+		// Contract
+		if (isModEnabled('contrat')) {
+			$langs->load('contracts');
+			$morehtmlref .=  '<br>';
+
+			$morehtmlref .=  $langs->trans('Contract');
+
+			if ($action != 'contrat') {
+				$morehtmlref .= '<a class="editfielda" href="'.$_SERVER["PHP_SELF"].'?action=contrat&amp;id='.$object->id.'">';
```
Comment (eldy) [inline htdocs/ticket/card.php:1012]: Can you rename the action = 'contrat'  into  'setcontract'  so CSRF security will be effective ?
PS: You will have to add also param token=newToken()...

File: htdocs/ticket/class/ticket.class.php [hunk 3]
```diff
@@ -35,6 +36,12 @@
  */
 class Ticket extends CommonObject
 {
+
+	/**
+	 * @var db connector
+	 */
+	public $bd;
```
Comment (frederic34) [inline htdocs/ticket/class/ticket.class.php:43]: $bd -> $db

File: htdocs/contrat/ticket.php [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+print '<div class="underbanner clearboth"></div>';
+
+
+/*
+ * Referers types
+ */
+
+$title=$langs->trans("ListTicketsLinkToContract");
+
+print '<table class="noborder" width="100%">';
```
Comment (frederic34) [inline htdocs/contrat/ticket.php:133]: noborder centpercent

File: htdocs/ticket/class/ticket.class.php [hunk 5]
```diff
@@ -2056,17 +2082,19 @@ public function setProgression($percent)
...
-		if ($result) {
-			$this->fk_contract = $contractid;
-			return 1;
+		if ($this->id) {
+			$sql = "UPDATE ".MAIN_DB_PREFIX."ticket";
+			$sql .= " SET fk_contract = ".($contractid > 0 ? $contractid : "null");
+			$sql .= " WHERE rowid = ".((int) $this->id);
+			dol_syslog(get_class($this).'::setContract sql='.$sql);
+			$resql = $this->db->query($sql);
+			print $sql;
```
Comment (frederic34) [inline htdocs/ticket/class/ticket.class.php:2091]: remove debug

File: htdocs/ticket/class/ticket.class.php [hunk 6]
```diff
@@ -86,7 +93,7 @@ class Ticket extends CommonObject
 	 * @var int Contract ID
 	 */
 	public $fk_contract;
-
+
```
Comment (frederic34) [inline htdocs/ticket/class/ticket.class.php:96]: superfluous tab

**Pull request merged**